sql/core/src/test/resources/sql-tests/results/cte.sql.out

```diff
@@ -328,16 +328,891 @@ struct<scalarsubquery():int>
+)
+SELECT COUNT(*) FROM r
+-- !query 46 schema
+struct<count(1):bigint>
+-- !query 46 output
+11
+
+
+-- !query 47
+WITH RECURSIVE r(level, data) AS (
+  VALUES (0, 'A')
+  UNION ALL
+  SELECT DISTINCT level + 1, data FROM r WHERE level < 10
+)
+SELECT * FROM r
+-- !query 47 schema
+struct<>
+-- !query 47 output
+org.apache.spark.sql.AnalysisException
+Recursive reference r cannot be used here. This can be caused by using it on inner side of an outer join, using it with aggregate or distinct, using it in a subquery or using it multiple times in a recursive term (except for using it on different sides of an UNION ALL).;
+
+
+-- !query 48
+WITH RECURSIVE y AS (
+  VALUES (1) AS t(id)
+),
+x AS (
+  SELECT * FROM y
+  UNION ALL
+  SELECT id + 1 FROM x WHERE id < 5
+)
+SELECT * FROM x
+-- !query 48 schema
+struct<id:int>
+-- !query 48 output
+1
+2
+3
+4
+5
+
+
+-- !query 49
+WITH RECURSIVE x AS (
+  VALUES (1) AS t(id)
+  UNION ALL
+  SELECT id + 1 FROM x WHERE id < 5
+),
+y AS (
+  VALUES (1) AS t(id)
+  UNION ALL
+  SELECT id + 1 FROM y WHERE id < 10
+)
+SELECT * FROM y LEFT JOIN x ON x.id = y.id
+-- !query 49 schema
+struct<id:int,id:int>
+-- !query 49 output
+1	1
+10	NULL
+2	2
+3	3
+4	4
+5	5
+6	NULL
+7	NULL
+8	NULL
+9	NULL
+
+
+-- !query 50
+WITH RECURSIVE x AS (
+  VALUES (1) AS t(id)
+  UNION ALL
+  SELECT id + 1 FROM x WHERE id < 5
+),
+y AS (
+  VALUES (1) AS t(id)
+  UNION ALL
+  SELECT id + 1 FROM x WHERE id < 10
+)
+SELECT * FROM y LEFT JOIN x ON x.id = y.id
+-- !query 50 schema
+struct<id:int,id:int>
+-- !query 50 output
+1	1
+2	2
+3	3
+4	4
+5	5
+6	NULL
+
+
+-- !query 51
+WITH RECURSIVE x AS (
+  SELECT 1 AS id
+  UNION ALL
+  SELECT id + 1 FROM x WHERE id < 3
+),
+y AS (
+  SELECT * FROM x
+  UNION ALL
+  SELECT * FROM x
+),
+z AS (
+  SELECT * FROM x
+  UNION ALL
+  SELECT id + 1 FROM z WHERE id < 10
+)
+SELECT * FROM z
+-- !query 51 schema
+struct<id:int>
+-- !query 51 output
+1
+10
+10
+10
+2
+2
+3
+3
+3
+4
+4
+4
+5
+5
+5
+6
+6
+6
+7
+7
+7
+8
+8
+8
+9
+9
+9
+
+
+-- !query 52
+WITH RECURSIVE x AS (
+  SELECT 1 AS id
+  UNION ALL
+  SELECT id + 1 FROM x WHERE id < 3
+),
+y AS (
+  SELECT * FROM x
+  UNION ALL
+  SELECT * FROM x
+),
+z AS (
+  SELECT * FROM y
+  UNION ALL
+  SELECT id + 1 FROM z WHERE id < 10
+)
+SELECT * FROM z
+-- !query 52 schema
+struct<id:int>
+-- !query 52 output
+1
+1
+10
+10
+10
+10
+10
+10
+2
+2
+2
+2
+3
+3
+3
+3
+3
+3
+4
+4
+4
+4
+4
+4
+5
+5
+5
+5
+5
+5
+6
+6
+6
+6
+6
+6
+7
+7
+7
+7
+7
+7
+8
+8
+8
+8
+8
+8
+9
+9
+9
+9
+9
+9
+
+
+-- !query 53
+WITH t AS (
+  WITH RECURSIVE s AS (
+    VALUES (1) AS t(i)
+    UNION ALL
+    SELECT i + 1 FROM s
+  )
+  SELECT i AS j FROM s LIMIT 10
+)
+SELECT * FROM t
+-- !query 53 schema
+struct<j:int>
+-- !query 53 output
+1
+10
+2
+3
+4
+5
+6
+7
+8
+9
+
+
+-- !query 54
+WITH RECURSIVE outermost AS (
+  WITH innermost AS (
+    SELECT * FROM outermost
+  )
+  SELECT level + 1 FROM innermost WHERE level < 5
+  UNION ALL
+  SELECT 0 AS level
+)
+SELECT * FROM outermost
```

Comment:
```
psql:with-recursive.sql:298: ERROR:  missing recursive reference
```

Reply:
I believe PostgreSQL doesn't allow this as 
```
WITH RECURSIVE outermost AS (
  WITH innermost AS (
    SELECT * FROM outermost
  )
  SELECT level + 1 FROM innermost WHERE level < 5
  UNION ALL
  SELECT 0 AS level
)
SELECT * FROM outermost
```
can be transformed to 
```
WITH RECURSIVE outermost AS (
  SELECT level + 1 FROM outermost WHERE level < 5
  UNION ALL
  SELECT 0 AS level
)
SELECT * FROM outermost
```
where we have the recursive and anchor terms in reverse order. My PR allows it.

Reply:
I think we need to clearly define which syntax is accepted or not, first.
Have you checked a statement definition for WITH RECURSIVE in the ANSI/SQL standard?
Then, could you define it in `SqlBase.sql`?
For example;
```
ctes
    : WITH namedQuery (',' namedQuery)*
    : WITH RECURSIVE name=errorCapturingIdentifier (columnAliases=identifierList)? AS? '(' recursiveQuery ')'
    ;
    
recursiveQuery
    : initStmt=queryPrimary UNION ALL? recursiveStmt=queryPrimary
```

Reply:
Although I agree with you on https://github.com/apache/spark/pull/23531#discussion_r307606058, I would not define the recursive structure in parsing rules in `SqlBase.g4`.
The reason is that `WITH` allows multiple CTEs to be defined and I would use the `RECURSIVE` flag to allow one or more of them to be recursive but not to force all of them to be recursive.
(Some of the relational DBs don't require the `RECURSIVE` flag to be there. I believe we should follow the PostgeSQL way, which means `RECURSIVE` flag is required to write recursive CTEs in a `WITH`, but doesn't mean that all CTEs have to be recursive.)
I also think we can give more sophisticated error messages if we handle recursion related issues in code like: https://github.com/apache/spark/pull/23531/files#diff-d0bfa3367c63988ad7cf33397e643e75R153-R156

Reply:
Yea, ok for now. But, at least, the check should be done in AstBuilder and it throws not AnalysisException but ParseException.

I spent some time to think about how to write code to support WITH RECURSIVE: https://github.com/maropu/spark/commit/fee3a19191d858c8adf9213f2653573882bd90fd

This code is not complete and the best though, I think some parts of them might be good references for you. Could you brush up your pr by checking my code?

Reply:
Cool, thanks @maropu! I will try to check and merge your changes into this PR.

Reply:
Thanks again! I like most of the changes and will incorporate them into this PR, but also left a few comments there.